eng/scripts/build.ps1

```diff
@@ -1,67 +1,74 @@
-$root = Resolve-Path ($PSScriptRoot + "/../..")
-Set-Location $root
-$sdks = @{};
-
-foreach ($sdk in (./eng/scripts/get_module_dirs.ps1 -serviceDir 'sdk/...')) {
-    $name = $sdk | split-path -leaf
-    $sdks[$name] = @{
-        'path'      = $sdk;
-        'clean'     = $clean;
-        'vet'       = $vet;
-        'generate'  = $generate;
-        'skipBuild' = $skipBuild;
-        'root'      = $root;
-    }
-}
-
-$keys = $sdks.Keys | Sort-Object;
-if (![string]::IsNullOrWhiteSpace($filter)) {
-    Write-Host "Using filter: $filter"
-    $keys = $keys.Where( { $_ -match $filter })
-}
 
-$keys | ForEach-Object { $sdks[$_] } | ForEach-Object {
-    Push-Location $_.path
-
-    if ($_.clean) {
-        Write-Host "##[command]Executing go clean -v ./... in " $_.path
+function Process-Sdk ($path) {
+    if ($clean) {
+        Write-Host "##[command]Executing go clean -v ./... in " $path
         go clean -v ./...
     }
 
-    if ($_.generate) {
-        Write-Host "##[command]Executing autorest.go in " $_.path
-        $autorestPath = $_.path + "/autorest.md"
+    if ($generate) {
+        Write-Host "##[command]Executing autorest.go in " $path
+        $autorestPath = $path + "/" + $config
 
         if (ShouldGenerate-AutorestConfig $autorestPath) {
             Generate-AutorestConfig $autorestPath
             $removeAutorestFile = $true
         }
 
         $autorestVersion = "@autorest/go@4.0.0-preview.23"
-        $outputFolder = $_.path
-        $root = $_.root
+        if ($outputFolder -eq '') {
+            $outputFolder = $path
+        }
         autorest --use=$autorestVersion --go --track2 --go-sdk-folder=$root --output-folder=$outputFolder --file-prefix="zz_generated_" --clear-output-folder=false $autorestPath
         if ($removeAutorestFile) {
             Remove-Item $autorestPath
         }
     }
-    if (!$_.skipBuild) {
-        Write-Host "##[command]Executing go build -v ./... in " $_.path
+
+    if (!$skipBuild) {
+        Write-Host "##[command]Executing go build -x -v ./... in " $path
         go build -x -v ./...
         Write-Host "##[command]Build Complete!"
 
     }
-    if ($_.vet) {
-        Write-Host "##[command]Executing go vet ./... in " $_.path
+
+    if ($vet) {
+        Write-Host "##[command]Executing go vet ./... in " $path
         go vet ./...
     }
-    Pop-Location
+
 }
 
-Set-Location $startingDirectory
+$startingDirectory = Get-Location
+$root = Resolve-Path ($PSScriptRoot + "/../..")
+Set-Location $root
+$sdks = @{};
+
+foreach ($sdk in (Get-ModuleDirs 'sdk/...')) {
+    $name = $sdk | split-path -leaf
+    $sdks[$name] = @{
+        'path' = $sdk;
+    }
+}
+
+$keys = $sdks.Keys | Sort-Object;
+if (![string]::IsNullOrWhiteSpace($filter)) {
+    Write-Host "Using filter: $filter"
+    $keys = $keys.Where( { $_ -match $filter })
+}
+
+try {
+    $keys | ForEach-Object { $sdks[$_] } | ForEach-Object {
+        Push-Location $_.path
+        Process-Sdk $_.path
+        Pop-Location
+    }
+}
```

Comment:
Now that I'm looking at my suggestion, I'm not convinced it's any better, but since I already wrote it I might as well show you.

If you default the filter value to `.*` then you can avoid all the extra data manipulation with one pipeline:

```
try {
    $sdks.GetEnumerator()
    | Where-Object {
      Write-Host "Using filter: $filter"
      $_.key -match $filter
    }
    | Sort-Object -Property key
    | ForEach-Object {
        Push-Location $_.value.path
        Process-Sdk $_.value.path
        Pop-Location
    }
}
```

Reply:
Hmm, I'm not sure I like yours better, but that is very interesting. I didn't realize you could do that in a one liner.